### PR TITLE
Add support for Squants.

### PR DESCRIPTION
**Pull request by cranst0n**

### CODE REVIEW

File: modules/squants/build.sbt
Comment (jcazevedo) [inline modules/squants/build.sbt:38]: You should add yourself as a developer here.

File: modules/squants/src/test/scala/pureconfig/module/squants/SquantsConvertTest.scala
Comment (leifwickland) [inline modules/squants/src/test/scala/pureconfig/module/squants/SquantsConvertTest.scala]: Nitpick: I think `tag.runtimeClass.getSimpleName` is a slightly more straightforward way to do this.

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (leifwickland) [inline modules/squants/src/main/scala/pureconfig/module/squants.scala]: Suggestion: 
What would you think about adding a case above this one something like:
```
case regex(_, unknownUnit) => Failure(new Exception(s"PureConfig library programmer error: The regex '$regex' found unit '$unknownUnit' which was not defined."))
```

I recognize this is intended to be temporary and is copied from squants, so feel free to disregard.

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (leifwickland) [inline modules/squants/src/main/scala/pureconfig/module/squants.scala]: Suggestion:
I'd be tempted to write this as the following to reduce duplication:
```
s match {
  case regex(value, unit) => unit match {
    case Fahrenheit.symbol | "f" | "F" => Success(Fahrenheit(value.toDouble))
    case Celsius.symbol | "c" | "F" => Success(Celsius(value.toDouble))
    // etc
  }
  case _ => Failure(QuantityParseException("Unable to parse Temperature", s))
}
```

I recognize this is intended to be temporary and is copied from squants, so feel free to disregard.

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (leifwickland) [inline modules/squants/src/main/scala/pureconfig/module/squants.scala]: Suggestion:
It feels a bit odd to me that multiple degree symbols are allowed.
I would replace `[ °]*` with ` *°? *` to only allow a single degree symbol.

I recognize this is intended to be temporary and is copied from squants, so feel free to disregard.

File: modules/squants/build.sbt
Comment (leifwickland) [inline modules/squants/build.sbt]: You ought to remove me. I don't deserve credit.

File: modules/squants/build.sbt
Comment (cranst0n) [reply modules/squants/build.sbt:38]: Added!

**Review by leifwickland — APPROVED** [review]
Awesome addition, @cranst0n!
Thank you!
I've made some suggestions. Take them or leave them.

**Comment by cranst0n** [comment]
Thanks @leifwickland. I've updated based on your feedback.

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (leifwickland) [inline modules/squants/src/main/scala/pureconfig/module/squants.scala]: @cranst0n I may have confused you. I was thinking that this error would only occur if `regex` included a unit character which wasn't supported here. I was hoping this wouldn't be an error that a normal user of the library would see.
The way you've defined things now, an input of `13°` will hit this case, reporting `°` as an invalid unit. `13(space)` will also hit this case.

I think you should go back to a defined set of expected units in the regex, causing this case to only be hit if the two lists of units end up diverging.

Comment (cranst0n) [reply modules/squants/src/main/scala/pureconfig/module/squants.scala]: @leifwickland Maybe I'm being thick here but I'm not sure exactly how you would get the behavior that you're looking for. If you match all supported units in the regex, something like
```scala
case regex(value, unit) => unit match {
  // cases covering F/C/K/R
}
case regex(_, unknownUnit) => ...
case _ => ...
```
The second top level case with `unknownUnit` will never hit. Or maybe I'm missing something on how scala's regex pattern matching works.

Comment (leifwickland) [reply modules/squants/src/main/scala/pureconfig/module/squants.scala]: @cranst0n 
This is me being an old paranoid engineer assuming that somebody somewhere is going to screw this up. 
This case that I'm describing would only be hit if somebody changed the regex but forgot to add a matching case for the unit.
The whole point is that the regex and the cases have diverged.
Since there's not a nice way to make the compiler check for that, I was adding a runtime check to give the pureconfig author a hint how we messed up.
Belt and suspenders I call it.
Sorry I confused you. Feel free to ignore my suggestion.

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (derekmorr) [reply modules/squants/src/main/scala/pureconfig/module/squants.scala]: Would you open an issue with squants for this? I'm a maintainer on squants and would be happy to look at a PR or issue.

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (derekmorr) [inline modules/squants/src/main/scala/pureconfig/module/squants.scala]: Would you open an issue with squants for this? I'm a maintainer on squants and would be happy to look at a PR or issue.

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (derekmorr) [reply modules/squants/src/main/scala/pureconfig/module/squants.scala]: Would you open an issue with squants for this? I'm a maintainer on squants and would be happy to look at a PR or issue.

**Comment by cranst0n** [comment]
@leifwickland I can appreciate a belt and suspenders approach. See my latest update. I've opened an issue here: typelevel/squants#189 for any further suggestions.

**Comment by derekmorr** [comment]
I opened a PR in squants to fix the double-degree issue: https://github.com/typelevel/squants/pull/190

I adapted the regex from the version here. I'm using:

```"([-+]?[0-9]*\\.?[0-9]+(?:[eE][-+]?[0-9]+)?)*[ ]*°? *(f|F|c|C|k|K|r|R)".r```

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (leifwickland) [inline modules/squants/src/main/scala/pureconfig/module/squants.scala]: Now that I have looked at squants a bit more, it occurs to me that there's no reason to have `Fahrenheit.symbol` etc in the cases because they can't match. They all include the degree symbol, which the regex will not include the capturing group.

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (leifwickland) [inline modules/squants/src/main/scala/pureconfig/module/squants.scala]: It looks like @derekmorr is updating this in squants. If you want to follow along now, it'd be:
```
"([-+]?[0-9]*\\.?[0-9]+(?:[eE][-+]?[0-9]+)?)* *°? *(f|F|c|C|k|K|r|R)".r
```

File: modules/squants/src/main/scala/pureconfig/module/squants.scala
Comment (derekmorr) [reply modules/squants/src/main/scala/pureconfig/module/squants.scala]: Right. This is the current version in squants - https://github.com/typelevel/squants/blob/6affa29e75c3d5109f8345e2263af9a24ee35fff/shared/src/main/scala/squants/thermal/Temperature.scala#L143 :

```scala
def apply(s: String): Try[Temperature] = {
    val regex = "([-+]?[0-9]*\\.?[0-9]+(?:[eE][-+]?[0-9]+)?)* *°? *(f|F|c|C|k|K|r|R)".r
    s match {
      case regex(value, unit) => unit match {
        case "f" | "F" => Success(Fahrenheit(value.toDouble))
        case "c" | "C" => Success(Celsius(value.toDouble))
        case "k" | "K" => Success(Kelvin(value.toDouble))
        case "r" | "R" => Success(Rankine(value.toDouble))
      }
      case _ => Failure(QuantityParseException("Unable to parse Temperature", s))
    }
  }
```

This version is probably going to be merged later today.

**Comment by derekmorr** [comment]
I don't know if you want to update the tests for this, but I added in a few more temperature-parsing tests into squants earlier today:

```scala
  they should "be flexible in parsing strings with regard to degree symbol and whitespace" in {
    Temperature("10.22 f").get should be (Fahrenheit(10.22))
    Temperature("10.22 °f").get should be (Fahrenheit(10.22))
    Temperature("10.22  °f").get should be (Fahrenheit(10.22))
    Temperature("10.22 ° f").get should be (Fahrenheit(10.22))
    Temperature("10.22 °   f").get should be (Fahrenheit(10.22))
    Temperature("10.22  °   f").get should be (Fahrenheit(10.22))
    Temperature("10.22°f").get should be (Fahrenheit(10.22))

    Temperature("10.22°°f").failed.get should be (QuantityParseException("Unable to parse Temperature", "10.22°°f"))
  }
```

Since you're duplicating the parsing code, it might be good to include the tests.

**Comment by melrief** [comment]
@leifwickland @jcazevedo thanks for having a look at this PR, I didn't have time. @leifwickland already approved the PR but I'll wait also for the approval from @jcazevedo as he commented on this PR too (just one thing about the authors, but I'll wait for LGTM anyway). I would like to merge this before #103 so I can merge this in my PR and switch also this module to the new result type.

**Review by melrief — APPROVED** [review]
Nothing to comment, everything is well done and I like how you wrote the tests. I'll publish an artifact for `0.5.1` right now.

**Comment by melrief** [comment]
I've published the package and announced it https://twitter.com/mapastr/status/830197666898444288

**Pull request merged**
